pkg/util/jsonmerge/jsonmerge.go

```diff
@@ -0,0 +1,115 @@
+package jsonmerge
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/evanphx/json-patch"
+	"github.com/golang/glog"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/yaml"
+)
+
+// Delta represents a change between two JSON documents.
+type Delta struct {
+	original []byte
+	edit     []byte
+}
+
+// NewDelta accepts two JSON or YAML documents and calculates the difference
+// between them.  It returns a Delta object which can be used to resolve
+// conflicts against a third version with a common parent, or an error
+// if either document is in error.
+func NewDelta(from, to []byte) (*Delta, error) {
+	d := &Delta{}
+	before, err := yaml.ToJSON(from)
+	if err != nil {
+		return nil, err
+	}
+	after, err := yaml.ToJSON(to)
+	if err != nil {
+		return nil, err
+	}
+	diff, err := jsonpatch.CreateMergePatch(before, after)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(6).Infof("Patch created from:\n%s\n%s\n%s", string(before), string(after), string(diff))
+	d.original = before
+	d.edit = diff
+	return d, nil
+}
+
+// Apply attempts to apply the changes described by Delta onto latest,
+// returning an error if the changes cannot be applied cleanly.
+// IsConflicting will be true if the changes overlap, otherwise a
+// generic error will be returned.
+func (d *Delta) Apply(latest []byte) ([]byte, error) {
```

Comment:
We can deal with anything JSON - the is definitely the low level structural tool.  We probably need to verify above it, perhaps by requiring a key match.

> On May 7, 2015, at 10:09 AM, David Eads <notifications@github.com> wrote:
> 
> In pkg/util/jsonmerge/jsonmerge.go:
> 
> > +	}
> > +	diff, err := jsonpatch.CreateMergePatch(before, after)
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	glog.V(6).Infof("Patch created from:\n%s\n%s\n%s", string(before), string(after), string(diff))
> > +	d.original = before
> > +	d.edit = diff
> > +	return d, nil
> > +}
> > +
> > +// Apply attempts to apply the changes described by Delta onto latest,
> > +// returning an error if the changes cannot be applied cleanly.
> > +// IsConflicting will be true if the changes overlap, otherwise a
> > +// generic error will be returned.
> > +func (d *Delta) Apply(latest []byte) ([]byte, error) {
> I think I'd like a layer of safety on this. Maybe an ApplyOnAPIObject? It's not a good idea to merge in changes if the API version has changed, even if those changes are non-conflicting.
> 
> —
> Reply to this email directly or view it on GitHub.
>